src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -0,0 +1,207 @@
+
+
+void MergeTreeBackgroundExecutor::wait()
+{
+    {
+        std::lock_guard lock(mutex);
+        shutdown = true;
+        has_tasks.notify_all();
+    }
+
+    pool.wait();
+}
+
+
+bool MergeTreeBackgroundExecutor::trySchedule(ExecutableTaskPtr task)
+{
+    std::lock_guard lock(mutex);
+
+    if (shutdown)
+        return false;
+
+    try
+    {
+        /// This is needed to increase / decrease the number of threads at runtime
+        /// Using stopwatch here not to do it so often.
+        /// No need to move the time to a config.
+        if (update_timer.compareAndRestartDeferred(10.))
+            updateConfiguration();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    auto & value = CurrentMetrics::values[metric];
+    if (value.load() >= static_cast<int64_t>(max_tasks_count))
+        return false;
+
+    pending.push_back(std::make_shared<TaskRuntimeData>(std::move(task), metric));
+
+    has_tasks.notify_one();
+    return true;
+}
+
+
+void MergeTreeBackgroundExecutor::removeTasksCorrespondingToStorage(StorageID id)
+{
+    std::vector<TaskRuntimeDataPtr> tasks_to_wait;
+    {
+        std::lock_guard lock(mutex);
+
+        /// Erase storage related tasks from pending and select active tasks to wait for
+        auto it = std::remove_if(pending.begin(), pending.end(),
+            [&] (auto item) -> bool { return item->task->getStorageID() == id; });
+        pending.erase(it, pending.end());
+
+        /// Copy items to wait for their completion
+        std::copy_if(active.begin(), active.end(), std::back_inserter(tasks_to_wait),
+            [&] (auto item) -> bool { return item->task->getStorageID() == id; });
+
+        for (auto & item : tasks_to_wait)
+            item->is_currently_deleting = true;
+    }
+
+
+    for (auto & item : tasks_to_wait)
+        item->is_done.wait();
+}
+
+
+void MergeTreeBackgroundExecutor::routine(TaskRuntimeDataPtr item)
+{
+    auto erase_from_active = [this, item]
+    {
+        active.erase(std::remove(active.begin(), active.end(), item), active.end());
+    };
+
+    try
+    {
+        if (item->task->executeStep())
+        {
+            std::lock_guard guard(mutex);
+
+            if (item->is_currently_deleting)
+            {
+                erase_from_active();
+                return;
+            }
+
+            pending.push_back(item);
+            erase_from_active();
+            has_tasks.notify_one();
+            return;
+        }
+
+        {
+            std::lock_guard guard(mutex);
+            erase_from_active();
+            has_tasks.notify_one();
+        }
+
+        /// In a situation of a lack of memory this method can throw an exception,
+        /// because it may interact somehow with BackgroundSchedulePool, which may allocate memory
+        /// But it is rather safe, because we have try...catch block here, and another one in ThreadPool.
+        item->task->onCompleted();
+        item->task.reset();
```

Comment:
As I see, here are just a few places where exception can happen: `task->executeStep()`, `task->onCompleted()` and `task.reset()`. I think it may be more logical to have a separate try/catch for the first and for the last two. Cause, even if we want to call `onCompleted` in case of exception in `task->executeStep()` (maybe have a special onException() for this?), we should not call `onCompleted` after exception in `onCompleted` or in `reset`.

Also maybe add a comment that exceptions from queues are considered to be not possible.

Reply:
Fixed.